File: packages/cli/src/commands/fetchPosts.ts
import * as log from '../utils/output/log';
import { gql, GraphQLClient } from 'graphql-request';
import { BaseArguments } from '../types/arguments';

interface QueryPostsArguments extends BaseArguments {
  limit: number;
  offset: number;
  json: boolean;
}

export const command = 'fetch_posts [limit] [offset] [json]';
export const desc =
  'Queries posts from the Oasis API. Returns an array of posts';

export const builder = {
  limit: {
    default: 10.0,
    describe: 'sets limit on the number of posts to query',
  },
  offset: {
    default: 0.0,
    describe:
      'offsets the posts that are being queries, useful for paginating posts',
  },
  json: {
    describe:
      'writes the raw JSON to stdout, powerful when used with jq (a JSON processor)',
  },
};

export const handler = async (yargs: QueryPostsArguments) => {
  const postsLimit = yargs.limit;
  const postsOffset = yargs.offset;
  const useJSON = yargs.json;

  const query = gql`
    query PaginatePosts($postsLimit: Float!, $postsOffset: Float!) {
      paginatePosts(limit: $postsLimit, offset: $postsOffset) {
        author {
          id
          name
          username
          badges {
            description
            id
            imagePath
          }
          avatar
        }
        createdAt
        downvotes
        id
        lastEdited
        upvotes
        message
        resort {
          id
          description
          logo
          name
        }
        comments(limit: 0, offset: 0) {
          total
        }
        topics
        isUpvoted
        isDownvoted
        imageName
      }
    }
  `;

  const client = new GraphQLClient(yargs.server, {
    headers: { authorization: yargs.auth },
  });

  const { paginatePosts } = await client.request(query, { postsLimit, postsOffset });

  if (useJSON) return console.log(JSON.stringify(paginatePosts));
  log.info(paginatePosts);
};

File: packages/tui/internal/queries/fetchPosts.go
package termoasis

import (
	"context"

	"github.com/shurcooL/graphql"
)

type PaginatedPosts struct {
	FeedSortPosts []struct {
		Message graphql.String
	} `graphql:"feedSortPosts(offset: $postsOffset, limit: $postsLimit)"`
}

func FetchPosts(client *graphql.Client, limit, offset float32) (
	// why, just why
	[]struct {
		Message graphql.String
		Author  struct {
			Id       graphql.String
			Name     graphql.String
			Username graphql.String
		}
		CreatedAt graphql.String
		Downvotes graphql.Int
		Upvotes   graphql.Int
	}, error) {
	var query struct {
		FeedSortPosts []struct {
			Message graphql.String
			Author  struct {
				Id       graphql.String
				Name     graphql.String
				Username graphql.String
			}
			CreatedAt graphql.String
			Downvotes graphql.Int
			Upvotes   graphql.Int
		} `graphql:"feedSortPosts(offset: $postsOffset, limit: $postsLimit)"`
	}

	err := client.Query(context.Background(), &query, map[string]interface{}{
		"postsOffset": graphql.Float(offset),
		"postsLimit":  graphql.Float(limit),
	})

	if err != nil {
		return nil, err
	}

	return query.FeedSortPosts, nil
}
